### PR TITLE
[ENH] MDNRegressor (Mixture Density Network)

### PR DESCRIPTION
**Pull request by joshdunnlime**
#### Reference Issues/PRs

No issue opened.

#### What does this implement/fix? Explain your changes.

A new regressor implementation of Mixture Density Network (MDN) as per [Bishop 1994](https://publications.aston.ac.uk/id/eprint/373/1/NCRG_94_004.pdf) with noise regularisation as per [Rothfuss 2019](https://arxiv.org/abs/1907.08982).
It also includes optional passing of pytorch activation functions and optimizers.

It implements a fully vectorised NormalMixture distribution where each rows weights are individually learnt and applied. It also implements a custom vectorised bisection for fast `_ppf` method calls.

#### Does your contribution introduce a new dependency? If yes, which one?

Yes, soft deps: [Pytorch](https://docs.pytorch.org/docs/stable/index.html) and [pytorch-optimizer](https://pytorch-optimizers.readthedocs.io/en/latest/).

#### What should a reviewer concentrate their feedback on?

NormalMixture distribution. This adds slightly opinionated design choice to the API.

#### Did you add any tests for the change?

Yes. Standard library param tests for dist and est. Additional test for est coming.

##### For all contributions
- [x] I've added myself to the [list of contributors](https://github.com/sktime/skpro/blob/main/CONTRIBUTORS.md) with any new badges I've earned :-)
  How to: add yourself to the [all-contributors file](https://github.com/sktime/skpro/blob/main/.all-contributorsrc) in the `skpro` root directory (not the `CONTRIBUTORS.md`). Common badges: `code` - fixing a bug, or adding code logic. `doc` - writing or improving documentation or docstrings. `bug` - reporting or diagnosing a bug (get this plus `code` if you also fixed the bug in the PR).`maintenance` - CI, test framework, release.
  See here for [full badge reference](https://allcontributors.org/docs/en/emoji-key)
- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

##### For new estimators
- [x] I've added the estimator to the API reference - in `docs/source/api_reference/taskname.rst`, follow the pattern.
- [x] I've added one or more illustrative usage examples to the docstring, in a pydocstyle compliant `Examples` section.
- [x] If the estimator relies on a soft dependency, I've set the `python_dependencies` tag and ensured
  dependency isolation, see the [estimator dependencies guide](https://www.sktime.net/en/latest/developer_guide/dependencies.html#adding-a-soft-dependency).

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -54,9 +54,11 @@ all_extras = [
     "mapie; python_version < '3.13'",
     "matplotlib>=3.3.2",
     "ngboost<0.6.0; python_version < '3.13'",
+    "pytorch-optimizer>=3.2.0; python_version < '3.14'",
```
Comment (fkiraly) [inline pyproject.toml]: I think for now we should not add dependencies to the `all_extras` depset - instead, let´s test the estimator on its own VM via the `tests:vm` tag set to `True`.

**Comment by joshdunnlime** [comment]
My understanding of `Mixture` is that it only has global weights for each sub-dist. MDNR _could_ use this, however, it would mean that for each prediction row, it would use a new instance of `Mixture` (computationally this would be crippling). `MixtureNormal` has performant, vectorised implementation of most of the class methods so is actually pretty quick!

You could argue that we could call it `Normal(Row|Instance|HeteroWeighted)Mixture` but I think it is documented clearly enough to avoid confusion.

**Comment by joshdunnlime** [comment]
I did consider the following:
1) Extending `Mixture` to include "instance" and "global" weights.
2) Create a generic `InstanceWeightedMixture` class.
3) A base class `BaseInstanceWeightedMixture`.

The downsides were:
1) Far too much work. IMHO these are two quite different classes to the point I think this would just be a wrapped around a `Mixture` and `InstanceWeightedMixture`.

2) `InstanceWeightedMixture` which accepts a mixture of different dists comes with a lot of complications. The primary one is that the `NormalMixture` has very well defined (and mostly exact) methods for all of the standard methods. My understanding of weighted mixtures of distributions stops a normals so I have no idea what is feasible here. The vectorisation is clearly very well defined for normal distributions in scipy, and assumptions on the support for root-finding in `_ppf` were fairly straightforward. Implemenation-wise, I believe it serves to keep `NormalMixture` even if we have 3) as the speed-ups and exactness are significant.

3) Happy to open a Issue. My knowledge of anything beyond normal mixtures of dists is effectively zero so I doubt I'd be up to a PR.

**Comment by joshdunnlime** [comment]
Just as a side note: The natural extension to MDNs is probably something Flow/Kernel based, as opposed to extending this with non-Normal dists.

File: docs/source/api_reference/regression.rst [hunk 2]
```diff
@@ -171,6 +171,17 @@ Distributional boosting
 
     XGBoostLSS
 
+Neural conditional density estimation
```
Comment (fkiraly) [inline docs/source/api_reference/regression.rst]: how about "deep learning based regressors" instead?

**Comment by joshdunnlime** [comment]
Thanks for updating. I'll do that optimizer package stuff.
I'm also adding some further improvements to the model which seem to help with training speed and stability.

**Comment by fkiraly** [comment]
btw, other topic, did we want to have the discussion on the differentiable transformers sometime? Have not seen you in the meetups, and I really think we need a sync meet on that (at least once). If the meetup times are not convenient, feel free to start a scheduling discussion on the discord.

**Comment by joshdunnlime** [comment]
I've added ngem loss: https://arxiv.org/html/2602.10602v1 to the bellow PR on my fork. It seems to show significant benefits at lower `lr`, particularly for CRPS. I am happy to merge into this `mdn` branch but further tidying is required. See comments in PR:
https://github.com/joshdunnlime/skpro/pull/1

It would be best to hold off and merge this all together to avoid minor breaking changes but I am unclear on how to handle the `ngem_lr` defaults as it really requires lower `lr` to perform well. Comments very much welcome.

Yes, apologies there. I will reach out on Discord regarding the Diff Transform class.
